communication/src/main/java/org/jboss/da/communication/pom/PomAnalyzer.java

```diff
@@ -0,0 +1,246 @@
+            }
+        } catch (TransferException | GalleyMavenException | URISyntaxException | CartoDataException
+                | IOException e) {
+            throw new PomAnalysisException(e);
+        }
+    }
+
+    /**
+     * Return an existing GAVDependencyTree if present in the map, otherwise create a new GAVDependencyTree
+     * and map the GAV to the GAVDependencyTree
+     *
+     * @param map
+     * @param gav
+     * @return
+     */
+    private GAVDependencyTree addGAVDependencyTreeToMapper(Map<GAV, GAVDependencyTree> map, GAV gav) {
+        return map.computeIfAbsent(gav, (k) -> new GAVDependencyTree(k));
+    }
+
+    private GAV generateGAV(ProjectVersionRef ref) {
+        return new GAV(ref.getGroupId(), ref.getArtifactId(), ref.getVersionString());
+    }
+
+    private DiscoveryResult getDiscoveryResult(File tempDir, File repoDir, ProjectVersionRef ref)
+            throws GalleyMavenException, URISyntaxException, CartoDataException {
+
+        MavenPomReader mavenPomReader = carto.getGalley().getPomReader();
+        MavenPomView pomView = mavenPomReader.read(ref, getRepoLocations(tempDir, repoDir));
+
+        Location location = new SimpleLocation("file:" + tempDir.getAbsolutePath());
+        URI src = new URI(location.getUri());
+
+        DiscoveryConfig disConf = new DiscoveryConfig(src);
+        disConf.setIncludeBuildSection(false);
+        disConf.setIncludeManagedDependencies(false);
+        disConf.setIncludeManagedPlugins(false);
+
+        MavenModelProcessor processor = new MavenModelProcessor();
+        return processor.readRelationships(pomView, src, disConf);
+    }
+
+    private List<Location> getRepoLocations(File cacheDir, File repoDir) {
+
+        Location location = new SimpleLocation("file:" + cacheDir.getAbsolutePath());
+
+        List<Location> repoLocations = new LinkedList<>();
+        // add local location
+        repoLocations.add(location);
+
+        // add maven central location
+        repoLocations.add(new SimpleLocation("central", "http://repo.maven.apache.org/maven2/"));
+
+        // add all the repositories mentioned in the pom.xml
+        List<SimpleLocation> locationsInPom = getLocationsDefinedInPom(repoDir);
+        repoLocations.addAll(locationsInPom);
+
+        return repoLocations;
+    }
+
+    private List<SimpleLocation> getLocationsDefinedInPom(File repoDir) {
+
+        List<String> pomFilePaths = findAllPomFiles(repoDir);
+        List<SimpleLocation> locations = new LinkedList<>();
+
+        for (String pom : pomFilePaths) {
+            Optional<MavenProject> project = pomReader.analyze(new File(pom));
+            if (project.isPresent() && project.get().getMavenRepositories() != null) {
+                for (MavenRepository repo : project.get().getMavenRepositories()) {
+                    locations.add(new SimpleLocation(repo.getId(), repo.getUrl()));
+                }
+            }
+        }
```

Comment:
```java
pomFilePaths.stream()
    .map(pom -> pomReader.analyze(new File(pom)))
    .forEach(project -> project
        .filter(x -> x.getMavenRepositories() != null)
        .ifPresent( x -> { x.getMavenRepositories().stream()
            .map( repo -> new SimpleLocation(repo.getId(), repo.getUrl()))
            .forEach( location -> locations.add(location))));
```
I :heart: lambda :goberserk: 

No, your way is good :smiley: . (When I see ``for`` I get an urge to rewrite it to lambda. Haven't seen a doctor about it yet.)